File: src/main/java/com/example/tutorial/dsa/medium/stackqueue/MinStackImplementation.java
package com.example.tutorial.dsa.medium.stackqueue;

import com.example.tutorial.common.datamodel.Offer;
import com.example.tutorial.common.utils.SampleDataLoader;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.stereotype.Component;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

/**
 * MinStackImplementation
 * ----------------------------------
 * <p>This program implements a stack that supports retrieving minimum element in O(1).
 * The core problem solved here is Min Stack (LeetCode #155).
 *
 * <p><b>Problem Statement:</b>
 * Design a stack that supports push, pop, top, and retrieving the minimum element
 * in constant time.
 *
 * <p><b>Real UseCase:</b> In a credit card offers system:
 * <ul>
 *   <li>Track minimum transaction amount in a processing queue</li>
 *   <li>Monitor lowest offer price in a dynamic list</li>
 *   <li>Keep track of minimum wait time in customer service queue</li>
 * </ul>
 *
 * <p><b>Operations:</b>
 * <ul>
 *   <li>push(val) - Push element onto stack</li>
 *   <li>pop() - Remove top element</li>
 *   <li>top() - Get top element</li>
 *   <li>getMin() - Get minimum element in O(1)</li>
 * </ul>
 *
 * <p><b>Company Tags:</b> Amazon, Microsoft, Bloomberg, Apple
 *
 * @see <a href="https://leetcode.com/problems/min-stack/">LeetCode 155 - Min Stack</a>
 */
@Component
public class MinStackImplementation implements CommandLineRunner {
  public static void main(String[] args) {
    SpringApplication.run(MinStackImplementation.class, args);
  }

  @Override
  public void run(String... args) throws Exception {
    System.out.println("=== MinStackImplementation: Min Stack Demo ===\n");

    // Load credit card offers from sample data
    List<Offer> offers = SampleDataLoader.OFFERS_DTO.get();
    System.out.println("Loaded " + offers.size() + " offers from sample data.\n");

    // Demonstrate MinStack with transaction amounts
    System.out.println("--- Tracking Minimum Transaction Amount ---\n");
    MinStack minStack = new MinStack();

    int[] transactions = {500, 200, 300, 100, 400};
    System.out.println("Processing transactions:");

    for (int amount : transactions) {
      minStack.push(amount);
      System.out.println("  Push " + amount + " → Min so far: " + minStack.getMin());
    }

    System.out.println("\nPopping elements:");
    while (!minStack.isEmpty()) {
      System.out.println("  Top: " + minStack.top() + ", Min: " + minStack.getMin());
      minStack.pop();
    }

    // Standard test case from LeetCode
    System.out.println("\n--- LeetCode Example ---\n");
    MinStack stack = new MinStack();
    stack.push(-2);
    System.out.println("push(-2)");
    stack.push(0);
    System.out.println("push(0)");
    stack.push(-3);
    System.out.println("push(-3)");
    System.out.println("getMin() → " + stack.getMin());  // -3
    stack.pop();
    System.out.println("pop()");
    System.out.println("top() → " + stack.top());        // 0
    System.out.println("getMin() → " + stack.getMin());  // -2
  }

  /**
   * MinStack - Stack with O(1) minimum retrieval.
   *
   * <p><b>DESIGN (Two-Stack Approach):</b>
   * <ol>
   *   <li>Main stack: stores all elements normally</li>
   *   <li>Min stack: stores minimum at each level</li>
   *   <li>When pushing: push to main, push min(val, currentMin) to minStack</li>
   *   <li>When popping: pop from both stacks</li>
   *   <li>getMin(): simply peek minStack</li>
   * </ol>
   *
   * <p><b>Example Walkthrough:</b>
   * <pre>
   * Operations: push(5), push(3), push(7), push(2), pop(), pop()
   *
   * push(5): mainStack=[5], minStack=[5]
   * push(3): mainStack=[5,3], minStack=[5,3] (3 < 5)
   * push(7): mainStack=[5,3,7], minStack=[5,3,3] (3 < 7)
   * push(2): mainStack=[5,3,7,2], minStack=[5,3,3,2] (2 < 3)
   *
   * getMin() → peek minStack → 2
   *
   * pop(): mainStack=[5,3,7], minStack=[5,3,3]
   * getMin() → 3
   *
   * pop(): mainStack=[5,3], minStack=[5,3]
   * getMin() → 3
   * </pre>
   *
   * <p><b>Time Complexity: O(1) for all operations</b>
   * <br>Push, pop, top, getMin are all constant time.
   * <br><i>Like having two stacks of plates - one regular, one tracking "smallest so far".
   * Every operation just looks at the top plate - instant!</i>
   *
   * <p><b>Space Complexity: O(n)</b>
   * <br>We use two stacks, each potentially holding n elements.
   * <br><i>Like keeping two identical-height stacks - double the storage,
   * but still proportional to number of elements.</i>
   */
  public static class MinStack {
    // Main stack for all elements.
    private final Deque<Integer> stack;

    // Parallel stack tracking minimum at each level.
    private final Deque<Integer> minStack;

    public MinStack() {
      stack = new ArrayDeque<>();
      minStack = new ArrayDeque<>();
    }

    /**
     * Pushes element onto the stack.
     *
     * @param val the value to push
     */
    public void push(int val) {
      stack.push(val);

      // Push current minimum to minStack.
      if (minStack.isEmpty()) {
        minStack.push(val);
      } else {
        minStack.push(Math.min(val, minStack.peek()));
      }
    }

    /**
     * Removes the top element from the stack.
     */
    public void pop() {
      if (!stack.isEmpty()) {
        stack.pop();
        minStack.pop();
      }
    }

    /**
     * Gets the top element.
     *
     * @return the top element
     */
    public int top() {
      return stack.peek();
    }

    /**
     * Retrieves the minimum element in the stack in O(1) time.
     *
     * @return the minimum element
     */
    public int getMin() {
      return minStack.peek();
    }

    /**
     * Checks if stack is empty.
     *
     * @return true if empty
     */
    public boolean isEmpty() {
      return stack.isEmpty();
    }
  }
}

File: src/main/python/com/example/tutorial/dsa/medium/stackqueue/min_stack_implementation.py
"""
MinStackImplementation
----------------------------------
This program implements a stack that supports retrieving minimum element in O(1).
The core problem solved here is Min Stack (LeetCode #155).

Problem Statement:
    Design a stack that supports push, pop, top, and retrieving the minimum element
    in constant time.

Real UseCase:
    In a credit card offers system:
    - Track minimum transaction amount in a processing queue
    - Monitor lowest offer price in a dynamic list
    - Keep track of minimum wait time in customer service queue

Operations:
    - push(val) - Push element onto stack
    - pop() - Remove top element
    - top() - Get top element
    - get_min() - Get minimum element in O(1)

Company Tags: Amazon, Microsoft, Bloomberg, Apple

See: https://leetcode.com/problems/min-stack/
"""

import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", ".."))

from com.example.tutorial.common.utils.sample_data_loader import load_offers


class MinStack:
    """
    MinStack - Stack with O(1) minimum retrieval.

    DESIGN (Two-Stack Approach):
        1. Main stack: stores all elements normally
        2. Min stack: stores minimum at each level
        3. When pushing: push to main, push min(val, current_min) to min_stack
        4. When popping: pop from both stacks
        5. get_min(): simply peek min_stack

    Example Walkthrough:
        Operations: push(5), push(3), push(7), push(2), pop(), pop()

        push(5): stack=[5], min_stack=[5]
        push(3): stack=[5,3], min_stack=[5,3] (3 < 5)
        push(7): stack=[5,3,7], min_stack=[5,3,3] (3 < 7)
        push(2): stack=[5,3,7,2], min_stack=[5,3,3,2] (2 < 3)

        get_min() -> peek min_stack -> 2

        pop(): stack=[5,3,7], min_stack=[5,3,3]
        get_min() -> 3

        pop(): stack=[5,3], min_stack=[5,3]
        get_min() -> 3

    Time Complexity: O(1) for all operations
        Push, pop, top, get_min are all constant time.
        Like having two stacks of plates - one regular, one tracking "smallest so far".
        Every operation just looks at the top plate - instant!

    Space Complexity: O(n)
        We use two stacks, each potentially holding n elements.
        Like keeping two identical-height stacks - double the storage,
        but still proportional to number of elements.
    """

    def __init__(self):
        """Initialize the MinStack."""
        # Main stack for all elements.
        self.stack = []

        # Parallel stack tracking minimum at each level.
        self.min_stack = []

    def push(self, val: int) -> None:
        """
        Pushes element onto the stack.

        Args:
            val: The value to push.
        """
        self.stack.append(val)

        # Push current minimum to min_stack.
        if not self.min_stack:
            self.min_stack.append(val)
        else:
            self.min_stack.append(min(val, self.min_stack[-1]))

    def pop(self) -> None:
        """Removes the top element from the stack."""
        if self.stack:
            self.stack.pop()
            self.min_stack.pop()

    def top(self) -> int:
        """
        Gets the top element.

        Returns:
            The top element.
        """
        return self.stack[-1]

    def get_min(self) -> int:
        """
        Retrieves the minimum element in the stack in O(1) time.

        Returns:
            The minimum element.
        """
        return self.min_stack[-1]

    def is_empty(self) -> bool:
        """
        Checks if stack is empty.

        Returns:
            True if empty.
        """
        return len(self.stack) == 0


def main():
    """Main function to demonstrate the MinStackImplementation."""
    print("=== MinStackImplementation: Min Stack Demo ===\n")

    # Load credit card offers from sample data
    offers = load_offers()
    print(f"Loaded {len(offers)} offers from sample data.\n")

    # Demonstrate MinStack with transaction amounts
    print("--- Tracking Minimum Transaction Amount ---\n")
    min_stack = MinStack()

    transactions = [500, 200, 300, 100, 400]
    print("Processing transactions:")

    for amount in transactions:
        min_stack.push(amount)
        print(f"  Push {amount} -> Min so far: {min_stack.get_min()}")

    print("\nPopping elements:")
    while not min_stack.is_empty():
        print(f"  Top: {min_stack.top()}, Min: {min_stack.get_min()}")
        min_stack.pop()

    # Standard test case from LeetCode
    print("\n--- LeetCode Example ---\n")
    stack = MinStack()
    stack.push(-2)
    print("push(-2)")
    stack.push(0)
    print("push(0)")
    stack.push(-3)
    print("push(-3)")
    print(f"get_min() -> {stack.get_min()}")  # -3
    stack.pop()
    print("pop()")
    print(f"top() -> {stack.top()}")          # 0
    print(f"get_min() -> {stack.get_min()}")  # -2


if __name__ == "__main__":
    main()
